src/System.Security.Claims/src/System/Security/Claims/ClaimsIdentity.cs

```diff
@@ -0,0 +1,970 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Diagnostics.Contracts;
+using System.Globalization;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Security.Principal;
+
+namespace System.Security.Claims
+{
+    /// <summary>
+    /// An Identity that is represented by a set of claims.
+    /// </summary>
+    [ComVisible(true)]
+    public class ClaimsIdentity : IIdentity
+    {
+        private enum SerializationMask
+        {
+            None = 0,
+            AuthenticationType = 1,
+            BootstrapConext = 2,
+            NameClaimType = 4,
+            RoleClaimType = 8,
+            HasClaims = 16,
+            HasLabel = 32,
+            Actor = 64,
+            UserData = 128,
+        }
+        private byte[] _userSerializationData;
+
+        private ClaimsIdentity _actor;
+        private string _authenticationType;
+        private object _bootstrapContext;
+        private Collection<IEnumerable<Claim>> _externalClaims = new Collection<IEnumerable<Claim>>();
```

Comment:
Could `_externalClaims` be typed as `List<List<Claim>>` instead of `Collection<IEnumerable<Claim>>`? It looks like this is only used internally and the `IEnumerable<Claim>`s are always `List<T>` already. This would be slightly more efficient than using `Collection<T>` and would avoid enumerator allocations in the `Claims` property when enumerating each item, as it could use `List<T>`'s struct enumerator.